### PR TITLE
fix: make jest ignore svg tags warnings

### LINKED ISSUE #50: 🤔 [ISSUE] - @eazychart-[react]- React unit tests are emitting warning about svg tag not recognized
### Description

React unit tests are emitting warning about svg tags not recognized by the browser. I believe that it has something to do with JEST that needs to be configured with the "dom" lib.

### Reproduction steps

```bash
1. `cd packages/ez-react`
2. Run `yarn test`
```

### Logs

```bash
PASS  tests/unit/components/shapes/Bar.spec.tsx
  ● Console

    console.error
      Warning: The tag <rect> is unrecognized in this browser. If you meant to render a React component, start its name with an uppercase letter.
          at rect
          at shapeDatum (/hdd/workspace/eazychart/packages/ez-react/src/components/shapes/Bar.tsx:13:3)
          at Legend (/hdd/workspace/eazychart/packages/ez-react/src/components/addons/legend/LegendProvider.tsx:10:3)
          at Tooltip (/hdd/workspace/eazychart/packages/ez-react/src/components/addons/tooltip/TooltipProvider.tsx:15:3)
          at dimensions (/hdd/workspace/eazychart/packages/ez-react/src/components/Chart.tsx:38:3)

      40 |     act(() => {
      41 |       // 1st render
    > 42 |       wrapper = render(
         |                 ^
      43 |         <Chart
      44 |           {...baseChartProps}
      45 |           scopedSlots={{

      at printWarning (node_modules/react-dom/cjs/react-dom.development.js:67:30)
      at error (node_modules/react-dom/cjs/react-dom.development.js:43:5)
      at createElement (node_modules/react-dom/cjs/react-dom.development.js:9027:9)
      at createInstance (node_modules/react-dom/cjs/react-dom.development.js:10192:20)
      at completeWork (node_modules/react-dom/cjs/react-dom.development.js:19464:28)
      at completeUnitOfWork (node_modules/react-dom/cjs/react-dom.development.js:22812:16)
      at performUnitOfWork (node_modules/react-dom/cjs/react-dom.development.js:22787:5)
      at workLoopSync (node_modules/react-dom/cjs/react-dom.development.js:22707:5)
      at renderRootSync (node_modules/react-dom/cjs/react-dom.development.js:22670:7)
      at performSyncWorkOnRoot (node_modules/react-dom/cjs/react-dom.development.js:22293:18)
      at scheduleUpdateOnFiber (node_modules/react-dom/cjs/react-dom.development.js:21881:7)
      at updateContainer (node_modules/react-dom/cjs/react-dom.development.js:25482:3)
      at node_modules/react-dom/cjs/react-dom.development.js:26021:7
      at unbatchedUpdates (node_modules/react-dom/cjs/react-dom.development.js:22431:12)
      at legacyRenderSubtreeIntoContainer (node_modules/react-dom/cjs/react-dom.development.js:26020:5)
      at Object.render (node_modules/react-dom/cjs/react-dom.development.js:26103:10)
      at node_modules/@testing-library/react/dist/pure.js:101:25
      at batchedUpdates$1 (node_modules/react-dom/cjs/react-dom.development.js:22380:12)
      at act (node_modules/react-dom/cjs/react-dom-test-utils.development.js:1042:14)
      at render (node_modules/@testing-library/react/dist/pure.js:97:26)
      at tests/unit/components/shapes/Bar.spec.tsx:42:17
      at batchedUpdates$1 (node_modules/react-dom/cjs/react-dom.development.js:22380:12)
      at act (node_modules/react-dom/cjs/react-dom-test-utils.development.js:1042:14)
      at _callee$ (tests/unit/components/shapes/Bar.spec.tsx:40:5)
      at tryCatch (node_modules/@babel/runtime/helpers/regeneratorRuntime.js:86:17)
      at Generator._invoke (node_modules/@babel/runtime/helpers/regeneratorRuntime.js:66:24)
      at Generator.next (node_modules/@babel/runtime/helpers/regeneratorRuntime.js:117:21)
      at asyncGeneratorStep (node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:24)
      at _next (node_modules/@babel/runtime/helpers/asyncToGenerator.js:25:9)
      at node_modules/@babel/runtime/helpers/asyncToGenerator.js:32:7
      at Object.<anonymous> (node_modules/@babel/runtime/helpers/asyncToGenerator.js:21:12)

```

### PR DESCRIPTION
**Pull request by G0maa**
# Motivation

Using `yarn test` in `/packages/ez-react/` yields in this specific warning `The tag <sometag> is unrecognized in this browser.`

Reference for solution: [here](https://stackoverflow.com/questions/58163226/is-it-possible-to-disable-specific-react-warnings-from-jest-using-create-react)

This solution ignores `console.error()` which has the text `is unrecognized in this browser.`, kind of hacky I guess.

Other solutions which I stumbled upon:
[This one](https://stackoverflow.com/questions/44467657/better-way-to-disable-console-inside-unit-tests) - Ignored because I don't know how to specify behavior depending on the error.

[This other one](https://stackoverflow.com/questions/70749049/how-do-i-suppress-a-specific-error-in-jest-tests) - Adds another dependency, also fails tests by default.

Fixes #50 

## Type of change

Please delete options that are not relevant.

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [X] I have done the work for both react and vue
- [X] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (Storybook)
- [ ] I have added unit tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

P.S: You got a really good configuration in mono-repo, git commit, CI pipeline : )

### CODE REVIEW

**Comment by marrouchi** [comment]
@G0maa Hello and thanks for your valuable contribution! Really awesome background check. The issue with this solution in my opinion is that we disable all the warning about dom inconsistency in react-dom : https://github.com/facebook/react/blob/2cf4352e1c81a5b8c3528519a128c20e8e65531d/packages/react-dom-bindings/src/client/ReactDOMComponent.js#L469 

After giving it a deep thought, I suggest to replace the console.error by a console.log, the same way you did. This will avoid the printing of the whole debug backtrace lines which are preventing us from having a proper jest display :
```
console.error = (...args) => console.warn('Error: ', ...args);
```

Let me know what you think :)

**Comment by G0maa** [comment]
Using this
```js
console.error = (...args) => console.log('Error: ', ...args);
```
Results in:
```
  ● Console

    console.log
      Warning: The tag <path> is unrecognized in this browser. If you meant to render a React component, start its name with an uppercase letter.
          at path
          at shapeData (D:\3_Studies\MyGithubFolder\3_OtherVentures\OpenSourceContirbution\eazychart\packages\ez-react\src\components\shapes\LinePath.tsx:14:3)
          at Legend (D:\3_Studies\MyGithubFolder\3_OtherVentures\OpenSourceContirbution\eazychart\packages\ez-react\src\components\addons\legend\LegendProvider.tsx:10:3)
          at Tooltip (D:\3_Studies\MyGithubFolder\3_OtherVentures\OpenSourceContirbution\eazychart\packages\ez-react\src\components\addons\tooltip\TooltipProvider.tsx:15:3)
          at dimensions (D:\3_Studies\MyGithubFolder\3_OtherVentures\OpenSourceContirbution\eazychart\packages\ez-react\src\components\Chart.tsx:38:3)

      at console.Object.<anonymous>.console.error (tests/ignoreSvgTags.ts:4:30)
```

and using this
```js
const ERROR = 'is unrecognized in this browser.';
console.error = (...args) => {
  if (args.toString().includes(ERROR))
    return console.log('Error: ', args[0].replace('%s', args[1]));
  else return console.log('Error: ', ...args);
};
```

Results in:
```
  ● Console

    console.log
      Error:  Warning: The tag <path> is unrecognized in this browser. If you meant to render a React component, start its name with an 
uppercase letter.%s

      at console.log [as error] (tests/ignoreSvgTags.ts:4:11)
```

This means that the `svg warning` gets somewhat minimized, while for all other warnings they'll get printed with less backtrace lines (as in the first example) in comparison to `console.warn()` or `console.log()`.


Pick you poison 👀

P.S: I'm not sure if the `return` in the second code example is necessary or not.

**Comment by G0maa** [comment]
@marrouchi I kind of need an answer 😬

**Comment by marrouchi** [comment]
@G0maa Sorry it took me so long to respond. After some digging I think we need to set an SVG container with the right namespaceURI  so that react-dom would bypass this check. This is just for some of the tests that are throwing the error. Here's an example : 
```js
it('renders an svg circle with the right coordinates', async () => {
    let wrapper: RenderResult;
    act(() => {
      const container = document.createElement('svg', {
        xmlns: 'http://www.w3.org/2000/svg',
      });
      wrapper = render(
        <Chart
          {...baseChartProps}
          scopedSlots={{
            LegendComponent: () => <>{null}</>,
            TooltipComponent: () => <>{null}</>,
          }}
        >
          <Point shapeDatum={pointA} />
        </Chart>,
        { container }
      );
    });
    await waitFor(() => {
      expect(wrapper.container.innerHTML).toMatchSnapshot();
    });
  });
```

To make it generic I suggest we create a custom function in tests/common.ts :
```js
export function renderSVG(
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
): RenderResult => {
    const container = document.createElement('svg', {
      xmlns: 'http://www.w3.org/2000/svg',
    });
   return render(ui, { ...options, container });
}
```

so that we have : 

```
it('renders an svg circle with the right coordinates', async () => {
    let wrapper: RenderResult;
    act(() => {
      wrapper = renderSVG(
        <Chart
          {...baseChartProps}
          scopedSlots={{
            LegendComponent: () => <>{null}</>,
            TooltipComponent: () => <>{null}</>,
          }}
        >
          <Point shapeDatum={pointA} />
        </Chart>
      );
    });
    await waitFor(() => {
      expect(wrapper.container.innerHTML).toMatchSnapshot();
    });
  });
```

**Comment by G0maa** [comment]
When I used `createElement` I got this `TS` error:
```
No overload matches this call.
  Overload 1 of 3, '(tagName: keyof HTMLElementTagNameMap, options?: ElementCreationOptions | undefined): HTMLObjectElement | ... 61 more ... | HTMLVideoElement', gave the following error.
    Argument of type '"svg"' is not assignable to parameter of type 'keyof HTMLElementTagNameMap'.
  Overload 2 of 3, '(tagName: keyof HTMLElementDeprecatedTagNameMap, options?: ElementCreationOptions | undefined): HTMLElement | ... 7 more ... | HTMLParamElement', gave the following error.
    Argument of type '"svg"' is not assignable to parameter of type 'keyof HTMLElementDeprecatedTagNameMap'.
  Overload 3 of 3, '(tagName: string, options?: ElementCreationOptions | undefined): HTMLElement', gave the following error.
    Argument of type '{ xmlns: string; }' is not assignable to parameter of type 'ElementCreationOptions'.
      Object literal may only specify known properties, and 'xmlns' does not exist in type 'ElementCreationOptions'.ts(2769)
```

I tried to use `createElementNS` which has `namespace URI` as specified in the [documentation](https://developer.mozilla.org/en-US/docs/Web/API/Document/createElementNS), but I got this other `TS` error instead:
```
Type 'RenderResult<typeof import("d:/3_Studies/MyGithubFolder/3_OtherVentures/OpenSourceContirbution/eazychart/packages/ez-react/node_modules/@testing-library/dom/types/queries"), SVGSVGElement, HTMLElement>' is not assignable to type 'RenderResult<typeof import("d:/3_Studies/MyGithubFolder/3_OtherVentures/OpenSourceContirbution/eazychart/packages/ez-react/node_modules/@testing-library/dom/types/queries"), HTMLElement, HTMLElement>'.
  Type 'RenderResult<typeof import("d:/3_Studies/MyGithubFolder/3_OtherVentures/OpenSourceContirbution/eazychart/packages/ez-react/node_modules/@testing-library/dom/types/queries"), SVGSVGElement, HTMLElement>' is not assignable to type '{ container: HTMLElement; baseElement: HTMLElement; debug: (baseElement?: Element | DocumentFragment | (Element | DocumentFragment)[] | undefined, maxLength?: number | undefined, options?: PrettyFormatOptions | undefined) => void; rerender: (ui: ReactElement<...>) => void; unmount: () => void; asFragment: () => Docu...'.
    Types of property 'container' are incompatible.
      Type 'SVGSVGElement' is missing the following properties from type 'HTMLElement': accessKey, accessKeyLabel, autocapitalize, dir, and 20 more.ts(2322)
  );
```
I'm not really good with `TS` so I'm not sure what to do.

**Comment by G0maa** [comment]
After some thinking I just edited `index.d.ts` to have:
```TS
  Container extends Element | DocumentFragment = HTMLElement | SVGSVGElement,
```
instead of just `HTMLElement`.

File: packages/ez-react/tests/common.ts [hunk 1]
```diff
@@ -16,3 +17,14 @@ export const baseChartProps: ChartProps = {
...
+export const renderSVG = (
+  ui: React.ReactElement,
+  options?: Omit<RenderOptions, 'queries'>
+): RenderResult => {
+  const container = document.createElementNS(
+    'http://www.w3.org/2000/svg',
+    'svg'
+  );
+  return render(ui, { ...options, container });
+};
```
Comment (marrouchi) [inline packages/ez-react/tests/common.ts:30]:
```suggestion

export const renderSVG = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>
): RenderResult => {
  const svg = document.createElementNS('http://www.w3.org/2000/svg', 'svg');
  return render(ui, {
    ...options,
    container: svg as Element as HTMLElement,
  });
};
```

Comment (marrouchi) [reply packages/ez-react/tests/common.ts:30]: @G0maa This should get rid of the TS error.

**Review by marrouchi — CHANGES_REQUESTED** [review]
Awesome work @G0maa :rocket: ! Just left you one comment about the TS issue. The update in index.d.ts is not included in the git versioning since it's under the node_modules folder.

File: packages/ez-react/tests/common.ts (hunk 1 above)
Comment (G0maa) [reply packages/ez-react/tests/common.ts]: Done! @marrouchi

**Review by marrouchi — APPROVED** [review]
Awesome work @G0maa ! :clap: 

Thank you for your contribution :guitar:

**Pull request merged**